api/logs/s3/s3.go

```diff
@@ -173,6 +175,245 @@ func (s *store) GetLog(ctx context.Context, appID, callID string) (io.Reader, er
+	})
+	if err != nil {
+		aerr, ok := err.(awserr.Error)
+		if ok && aerr.Code() == s3.ErrCodeNoSuchKey {
+			return nil, models.ErrCallNotFound
+		}
+		return nil, fmt.Errorf("failed to read log, %v", err)
+	}
+
+	var call models.Call
+	err = json.Unmarshal(target.Bytes(), &call)
+	if err != nil {
+		return nil, err
+	}
+
+	return &call, nil
+}
+
+func flipCursor(oid string) string {
+	if oid == "" {
+		return ""
+	}
+
+	return id.EncodeDescending(oid)
+}
+
+func callMarkerKey(app, path, id string) string {
+	id = flipCursor(id)
+	// s3 urls use / and are url, we need to encode this since paths have / in them
+	path = base64.RawURLEncoding.EncodeToString([]byte(path))
+	return callMarkerPrefix + app + "/" + path + "/" + id
+}
+
+func callKey(app, id string) string {
+	id = flipCursor(id)
+	return callKeyFlipped(app, id)
+}
+
+func callKeyFlipped(app, id string) string {
+	return callKeyPrefix + app + "/" + id
+}
+
+func logKey(appID, callID string) string {
+	return logKeyPrefix + appID + "/" + callID
+}
+
+// GetCalls returns a list of calls that satisfy the given CallFilter. If no
+// calls exist, an empty list and a nil error are returned.
+// NOTE: this relies on call ids being lexicographically sortable and <= 16 byte
+func (s *store) GetCalls(ctx context.Context, filter *models.CallFilter) ([]*models.Call, error) {
+	ctx, span := trace.StartSpan(ctx, "s3_get_calls")
+	defer span.End()
+
+	if filter.AppID == "" {
+		return nil, errors.New("s3 store does not support listing across all apps")
+	}
+
+	// NOTE:
+	// if filter.Path != ""
+	//   find marker from marker keys, start there, list keys, get next marker from there
+	// else
+	//   use marker for keys
+
+	// NOTE we need marker keys to support (app is REQUIRED):
+	// 1) quick iteration per path
+	// 2) sorted by id across all path
+	// marker key: m : {app} : {path} : {id}
+	// key: s: {app} : {id}
+	//
+	// also s3 api returns sorted in lexicographic order, we need the reverse of this.
+
+	// marker is either a provided marker, or a key we create based on parameters
+	// that contains app_id, may be a marker key if path is provided, and may
+	// have a time guesstimate if to time is provided.
+
+	var marker string
+
+	// filter.Cursor is a call id, translate to our key format. if a path is
+	// provided, we list keys from markers instead.
+	if filter.Cursor != "" {
+		marker = callKey(filter.AppID, filter.Cursor)
+		if filter.Path != "" {
+			marker = callMarkerKey(filter.AppID, filter.Path, filter.Cursor)
+		}
+	} else if t := time.Time(filter.ToTime); !t.IsZero() {
+		// we need to use first 48 bits of id to approximate created_at, to skip over records
+		// if looking far into the past to save time.
+		ms := uint64(t.Unix()*1000) + uint64(t.Nanosecond()/int(time.Millisecond))
+		var buf id.Id
+		buf[0] = byte(ms >> 40)
+		buf[1] = byte(ms >> 32)
+		buf[2] = byte(ms >> 24)
+		buf[3] = byte(ms >> 16)
+		buf[4] = byte(ms >> 8)
+		buf[5] = byte(ms)
+		mid := buf.String()[:10]
```

Comment:
should we call buf.MarshalTextTo() to translate this into base32 first?

Reply:
not sure i'm tracking, the string method will call that. we could use an array and save an alloc though, can do that.

Reply:
Cool, I missed that.. :-)